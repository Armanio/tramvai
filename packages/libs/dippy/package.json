{
  "name": "@tinkoff/dippy",
  "version": "0.0.0-stub",
  "initialVersion": "0.7.27",
  "description": "",
  "main": "./lib/di.js",
  "typings": "./lib/di.d.ts",
  "files": [
    "__migrations__",
    "lib"
  ],
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "git@github.com:TinkoffCreditSystems/tramvai.git"
  },
  "scripts": {
    "build": "tramvai-build",
    "watch": "tsc -w",
    "build-for-publish": "true"
  },
  "gitHead": "8e826a214c87b188fc4d254cdd8f2a2b2c55f3a8",
  "dependencies": {
    "tslib": "^2.0.3"
  },
  "module": "./lib/di.es.js",
  "es2017": "./lib/di.es2017.js",
  "license": "Apache-2.0"
}
