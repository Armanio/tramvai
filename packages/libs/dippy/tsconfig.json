{
  "extends": "../../../tsconfig.lib.json",
  "include": [
    "./src"
  ],
  "compilerOptions": {
    "module": "CommonJS",
    "declarationDir": "./lib",
    "outDir": "./lib",
    "rootDir": "./src",
    "strict": false,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "strictBindCallApply": false,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true
  }
}
