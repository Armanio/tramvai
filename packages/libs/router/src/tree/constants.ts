export const PARAMETER_DELIMITER = ':';
export const WILDCARD_REGEXP = /\*/;
export const HISTORY_FALLBACK_REGEXP = /<history-fallback>/;
export const PARAM_PARSER_REGEXP = /([^(?]+)(?:\((.+)\))?(\?)?/;
