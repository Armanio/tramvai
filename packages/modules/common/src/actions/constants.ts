export const actionType = {
  global: 'global' as const,
  local: 'local' as const,
};

export type ActionType = keyof typeof actionType;
