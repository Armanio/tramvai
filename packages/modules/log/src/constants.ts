export const LOGGER_NAME = '';
export const LOGGER_KEY = 'tramvai';
export const LOGGER_DEPTH_LIMIT = 20;
