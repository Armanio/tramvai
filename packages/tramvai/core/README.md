# Core

Ядро tramvai. В основном требуется для разработки модулей трамвая.

## Экспортируемые токены

### DI_TOKEN

Используется для доступа к инстансу DI

### APP_INFO_TOKEN

Содержит информацию о запущенном приложении

### COMMAND_LINE_RUNNER_TOKEN

Реализация раннера для исполнения команд

### COMMAND_LINES_TOKEN

Задание списка команд
