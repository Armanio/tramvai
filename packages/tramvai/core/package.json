{
  "name": "@tramvai/core",
  "version": "0.0.0-stub",
  "description": "",
  "main": "lib/index.js",
  "typings": "src/index.ts",
  "files": [
    "lib",
    "__migrations__"
  ],
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "git@github.com:TinkoffCreditSystems/tramvai.git"
  },
  "scripts": {
    "build": "tramvai-build",
    "watch": "tsc -w",
    "build-for-publish": "true"
  },
  "dependencies": {
    "@tinkoff/utils": "^2.1.2",
    "@tramvai/types-actions-state-context": "0.0.0-stub",
    "@tramvai/tools-check-versions": "0.0.0-stub",
    "@tramvai/tools-migrate": "0.0.0-stub",
    "@tinkoff/dippy": "0.0.0-stub",
    "tslib": "^2.0.3"
  },
  "peerDependencies": {
    "jscodeshift": "^0.7.0"
  },
  "gitHead": "8e826a214c87b188fc4d254cdd8f2a2b2c55f3a8",
  "module": "lib/index.es.js",
  "es2017": "lib/index.es2017.js",
  "license": "Apache-2.0"
}
