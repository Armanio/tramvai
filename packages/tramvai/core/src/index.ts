export { createApp, App } from './createApp';
export { createBundle } from './bundles/createBundle';
export { Bundle, BundleImport } from './bundles/createBundle.h';
export { createAction } from './actions/createActions';
export * from './types/action';
export { Module, module, MODULE_PARAMETERS, provide } from './modules/module';
export { getModuleParameters } from './modules/getModuleParameters';
export { walkOfModules } from './modules/walkOfModules';
export { isExtendedModule } from './modules/isExtendedModule';
export { ModuleType, ExtendedModule } from './modules/module.h';
export { Command } from './command/command.h';
export * from './tokens';

export { Scope, Provider } from '@tinkoff/dippy';
