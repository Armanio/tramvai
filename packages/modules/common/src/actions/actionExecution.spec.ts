import { createAction, ACTION_PARAMETERS } from '@tramvai/core';
import { createContainer } from '@tinkoff/dippy';
import { ActionExecution } from './actionExecution';
import { alwaysCondition } from './conditions/always';
import { actionType } from './constants';

const contextMock = { isContext: true };

const createActionNameFactory = (name, result, conditions = {}) =>
  createAction({
    fn: (c, payload) => result.push(`action: ${name} with: ${payload}`),
    name,
    conditions,
  });

describe('actionExecution in global page', () => {
  it('Запуск page действий без каких либо опций', async () => {
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
    });
    const result = [];

    const action1 = createActionNameFactory('action1', result);
    const action2 = createActionNameFactory('action2', result);

    await instance.run(action1, '@globalAction', actionType.global);
    await instance.run(action2, '@globalAction', actionType.global);
    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action1 with: @globalAction",
  "action: action2 with: @globalAction",
]
`);
    // Экшен не должен выполниться. Автоматическиая дедубликация
    await instance.run(action1, '@globalAction', actionType.global);
    await instance.run(action1, '@globalAction', actionType.global);
    await instance.run(action2, '@globalAction', actionType.global);
    await instance.run(action2, '@globalAction', actionType.global);
    await instance.run(action1, '@globalAction', actionType.global);

    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action1 with: @globalAction",
  "action: action2 with: @globalAction",
]
`);
  });

  it('Запуск page экшенов с фильтрами', async () => {
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [
        {
          key: 'neverCondition',
          fn: (cheker) => {
            if (cheker.conditions.never) {
              cheker.forbid();
            }
          },
        },
      ],
      // @ts-ignore
      context: contextMock,
    });
    const result = [];

    await instance.run(
      createActionNameFactory('action1', result),
      '@globalAction',
      actionType.global
    );
    await instance.run(
      createActionNameFactory('action2', result, { never: true }),
      '@globalAction',
      actionType.global
    );
    await instance.run(
      createActionNameFactory('action3', result, { never: true }),
      '@globalAction',
      actionType.global
    );
    await instance.run(
      createActionNameFactory('action4', result),
      '@globalAction',
      actionType.global
    );
    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action1 with: @globalAction",
  "action: action4 with: @globalAction",
]
`);
  });

  it('Множественный запуск page actions с проверкой, что только измененые экшены запускаются ', async () => {
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [alwaysCondition],
      // @ts-ignore
      context: contextMock,
    });
    const result = [];

    const action1 = createActionNameFactory('action multi 1', result);
    const action2 = createActionNameFactory('action multi 2', result, { always: true });

    await instance.run(action1, '@globalAction', actionType.global);
    await instance.run(action2, '@globalAction', actionType.global);

    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action multi 1 with: @globalAction",
  "action: action multi 2 with: @globalAction",
]
`);

    await instance.run(
      createActionNameFactory('action multi 1', result),
      '@globalAction',
      actionType.global
    );
    await instance.run(action2, '@globalAction', actionType.global);

    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action multi 1 with: @globalAction",
  "action: action multi 2 with: @globalAction",
  "action: action multi 2 with: @globalAction",
]
`);
  });

  it('Востанавливаем состояние выполненных экшенов и проверяем на влияние будущих экшенов', async () => {
    const state = {
      actionTramvai: {
        serverState: {
          'action resume 1': { status: 'success', state: { test: '@globalAction' } },
          'action resume 3': { status: 'success', state: { test: '@globalAction' } },
        },
      },
    };
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return state[reducer.storeName];
          }
          return state;
        },
      } as any,
      actionConditionals: [
        {
          key: 'test',
          fn: (cheker) => {
            const last = cheker.payload;
            if (cheker.getState() !== last) {
              cheker.setState(last);
              cheker.allow();
            }
          },
        },
      ],
      // @ts-ignore
      context: contextMock,
    });
    const result = [];

    const action1 = createActionNameFactory('action resume 1', result);
    const action2 = createActionNameFactory('action resume 2', result);
    const action3 = createActionNameFactory('action resume 3', result);

    await instance.run(action1, '@globalAction', actionType.global);
    await instance.run(action2, '@globalAction', actionType.global);
    await instance.run(action3, '@globalAction', actionType.global);
    await instance.run(action3, '@globalAction', actionType.global);
    await instance.run(action3, '@globalAction', actionType.global);

    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action resume 2 with: @globalAction",
]
`);
  });

  it('Получение и использование значений переданных в conditions экшенов', async () => {
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [
        {
          key: 'value',
          fn: (cheker) => {
            const { role } = cheker.conditions;
            if (role) {
              if (cheker.getState() !== role) {
                cheker.setState(role);
                cheker.allow();
              }
            }
          },
        },
      ],
      // @ts-ignore
      context: contextMock,
    });
    const result = [];

    const action1 = createActionNameFactory('action 1', result, { role: 'client' });
    const action2 = createActionNameFactory('action 1', result, { role: 'admin' });

    await instance.run(action1, '@globalAction', actionType.global);
    await instance.run(action1, '@globalAction', actionType.global);
    await instance.run(action1, '@globalAction', actionType.global);
    await instance.run(action2, '@globalAction', actionType.global);
    await instance.run(action2, '@globalAction', actionType.global);
    await instance.run(action1, '@globalAction', actionType.global);
    // Выполняем экшены по одному разу, дальше из-за одинакового conditions, скипаем

    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action 1 with: @globalAction",
  "action: action 1 with: @globalAction",
  "action: action 1 with: @globalAction",
]
`);
  });

  it('Передаем в экшены di, если были добавлены в deps', async () => {
    const instance = new ActionExecution({
      di: createContainer([
        { provide: 'dep1', useValue: 'dep1 value' },
        { provide: 'dep2', useValue: 'dep2 value' },
      ]),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
    });

    const result = [];
    await instance.run(
      createAction({
        fn: (context, payload, deps) => result.push([context, payload, deps]),
        name: 'action1',
        deps: { dep1: 'dep1', dep2: 'dep2' },
      }),
      '@globalAction',
      actionType.global
    );
    await instance.run(
      createAction({
        fn: (context, payload, deps) => result.push([context, payload, deps]),
        name: 'action2',
        deps: { dep2: 'dep2' },
      }),
      '@globalAction',
      actionType.global
    );
    // Экшен не должен выполниться. Автоматическиая дедубликация
    await instance.run(
      createAction({
        fn: (context, payload, deps) => result.push([context, payload, deps]),
        name: 'action1',
      }),
      '@globalAction',
      actionType.global
    );

    expect(result).toMatchSnapshot();
  });

  it('Преобразование экшенов с старого вида в новый через transformAction', async () => {
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
      transformAction: (action: any) => {
        if (!action[ACTION_PARAMETERS]) {
          return createAction({ fn: action, name: action.name });
        }
        return action;
      },
    });
    const result = [];
    await instance.run(
      createActionNameFactory('action1', result),
      '@globalAction',
      actionType.global
    );
    const actionOld = (c, payload) => result.push(payload);
    await instance.run(actionOld as any, '@globalAction', actionType.global);

    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action1 with: @globalAction",
  "@globalAction",
]
`);
  });
});

describe('actionExecution in actions', () => {
  it('Запуск множество экшенов без каких либо опций с изначальным стейтом', async () => {
    const state = {
      actionTramvai: {
        serverState: {
          action1: { status: 'success', state: {} },
          action2: { status: 'success', state: {} },
        },
      },
    };
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return state[reducer.storeName];
          }
          return state;
        },
      } as any,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
    });
    const result = [];

    const action1 = createActionNameFactory('action1', result);
    const action2 = createActionNameFactory('action2', result);

    await instance.run(action1, 1);
    await instance.run(action2, 2);
    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action1 with: 1",
  "action: action2 with: 2",
]
`);
    // Экшен не должен выполниться. Автоматическиая дедубликация
    await instance.run(action1, 3);
    await instance.run(action1, 4);
    await instance.run(action2, 5);

    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action1 with: 1",
  "action: action2 with: 2",
  "action: action1 with: 3",
  "action: action1 with: 4",
  "action: action2 with: 5",
]
`);
  });

  it('Ограничение по ролям', async () => {
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [
        {
          key: 'roles',
          fn: (cheker) => {
            const { role } = cheker.conditions;
            if (role) {
              if (role !== 'admin') {
                cheker.forbid();
              }
            }
          },
        },
      ],
      // @ts-ignore
      context: contextMock,
    });
    const result = [];

    const action1 = createActionNameFactory('action 1', result, { role: 'client' });
    const action2 = createActionNameFactory('action 2', result, { role: 'admin' });

    await instance.run(action1, 1);
    await instance.run(action2, 2);
    await instance.run(action2, 3);
    await instance.run(action1, 4);
    // Выполняем экшены по одному разу, дальше из-за одинакового conditions, скипаем

    expect(result).toMatchInlineSnapshot(`
Array [
  "action: action 2 with: 2",
  "action: action 2 with: 3",
]
`);
  });
});
