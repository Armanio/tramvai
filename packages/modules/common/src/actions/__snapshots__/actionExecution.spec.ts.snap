// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actionExecution in global page Передаем в экшены di, если были добавлены в deps 1`] = `
Array [
  Array [
    Object {
      "isContext": true,
    },
    "@globalAction",
    Object {
      "dep1": "dep1 value",
      "dep2": "dep2 value",
    },
  ],
  Array [
    Object {
      "isContext": true,
    },
    "@globalAction",
    Object {
      "dep2": "dep2 value",
    },
  ],
]
`;
